flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/inference/TypeInferenceOperandChecker.java

```diff
@@ -240,10 +246,28 @@ private void insertImplicitCasts(SqlCallBinding callBinding, List<DataType> expe
 
     /** Adopted from {@link org.apache.calcite.sql.validate.implicit.AbstractTypeCoercion}. */
     private SqlNode castTo(SqlNode node, RelDataType type) {
-        return SqlStdOperatorTable.CAST.createCall(
-                SqlParserPos.ZERO,
-                node,
-                SqlTypeUtil.convertTypeToSpec(type).withNullable(type.isNullable()));
+        final SqlDataTypeSpec dataType;
+        if (type instanceof RawRelDataType) {
+            dataType = createRawDataTypeSpec((RawRelDataType) type);
+        } else {
+            dataType = SqlTypeUtil.convertTypeToSpec(type).withNullable(type.isNullable());
+        }
+
+        return SqlStdOperatorTable.CAST.createCall(SqlParserPos.ZERO, node, dataType);
+    }
+
+    private SqlDataTypeSpec createRawDataTypeSpec(RawRelDataType type) {
+        final RawType<?> rawType = type.getRawType();
+
+        SqlNode className =
+                SqlLiteral.createCharString(
+                        rawType.getOriginatingClass().getName(), SqlParserPos.ZERO);
+        SqlNode serializer =
+                SqlLiteral.createCharString(rawType.getSerializerString(), SqlParserPos.ZERO);
+
+        SqlTypeNameSpec rawSpec = new SqlRawTypeNameSpec(className, serializer, SqlParserPos.ZERO);
+
+        return new SqlDataTypeSpec(rawSpec, null, type.isNullable(), SqlParserPos.ZERO);
```

Comment:
can we have test for that?

Reply:
A test in stream.FunctionITCase should be enough.